templates/fine-tune-function-calling/fc_utils/function_extraction_utils.py

```diff
@@ -0,0 +1,65 @@
+"""
+Function call json extraction utils.
+Reference: https://gist.github.com/kouroshHakha/6cfbe2bf4aaafc5db733d408044f9902#file-create_test_dataset-py
+"""
+
+import json
+import re
+from fc_utils.preprocessing import TOOL_CALL_TAGS, TOOL_RESULT_TAGS
+
+class FunctionCallNotFoundError(Exception):
+    pass
+
+class FunctionResponseNotFoundError(Exception):
+    pass
+
+def extract_jsons(jsonl_string):
+    """
+    Extracts JSON objects from a string containing one or more JSONs.
+    Example: [{"name": "weather", "arguments": {\"location\": \"New York\"}}]
+    """
+    json_strs = []
+    brace_count = 0
+    start = None
+
+    for i, char in enumerate(jsonl_string):
+        if char == '{':
+            if brace_count == 0:
+                start = i
+            brace_count += 1
+        elif char == '}':
+            brace_count -= 1
+            if brace_count == 0:
+                end = i + 1
+                json_strs.append(jsonl_string[start:end])
```

Comment:
use regex or sth less adhoc and more systematic?

Reply:
Ask AI to improve your code.

Reply:
Is this not systematic? i keep track of open braces and then when all are closed, I extract json from the first open brace. It will work with all sorts of special characters etc. 


Regex was my go-to but I've encountered issues with nested JSONs. The format here is that  you have a string which contains one or more jsons separated by commas. You can have nested jsons separated by commas and regex won't work then (atleast the patterns I tried).  The code here is a bit slower but will be correct.

Reply:
This code is not maintainable. You are adding complexity in a place whose concern is not parsing json. So the preference is to find a one / two liner usining an existing popular library to do the parsing. Say tomorrow this part of the code turns out to be problematic, it will be a lot harder for someone to resolve the issue then. So my question to you is what is the complex jsonl_string that you couldn't use regex for?

Reply:
A part of the problem was that I was using the same function in some places to process Glaive's system message (where the jsons are space separated) and our formatted tool calls (where they are comma separated). I've refactored the whole thing now to have use to separate the two kinds of formats. I've also used one common function to process our tool info formatting (which is "[prefix_content] <tag1> [tool info] <tag2>") for tool lists, tool calls, tool results, etc. and then leave further parsing to `json`.